### PR TITLE
New VersionGenerator for the Hot Rod protocol

### LINKED ISSUE #13191: Create a new version generator for Hot Rod writes
### Describe the bug

The current version generator for the Hot Rod writes has the following format:

`| view id (2 bytes) | rank (2 bytes) | version counter (4 bytes) |`

It works fine for single cluster deployments but not for cross-site deployments. When cross-site is enabled, the `view id` and `rank` may match and generate the same version on both sites. 

For example, a single node on 2 sites will have the same `view id` (it will be 1) and `rank` (once again 1 on both sites since they are the only member).

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Causes https://github.com/keycloak/keycloak/issues/35015

### PR DESCRIPTION
**Pull request by pruivo**
Closes  #13191

### CODE REVIEW

**Comment by pruivo** [comment]
Alternatives I considered

### New Implementation using UUID instead of Random
Using UUID and xor the most and least significant bits
```java
var uuid = UUID.randomUUID();
new NumericVersion(Math.abs(uuid.getMostSignificantBits() ^ uuid.getLeastSignificantBits()));
```
It seems more expensive and generates more garbage. 

### Modify `RankCalculator`

`RankCalculator`  computes the prefix using the `viewId` and `index` and I thought about using the `Address.hashCode`. The JGroups `Address` is a UUID-based one with the following hash code computation: `this.mostSigBits >> 32 ^ this.mostSigBits ^ this.leastSigBits >> 32 ^ this.leastSigBits`. It would be enough to ensure each node had a different prefix. 

I don't have a good reason to discard this one. 

Let me know your preference.

**Comment by jabolina** [comment]
I'm going to take a look at the code next. But I'm +1 on having the version take into account the node's address instead of being random.

File: core/src/main/java/org/infinispan/container/versioning/RandomVersionGenerator.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+   private RandomVersionGenerator() {}
+
+   @Override
+   public NumericVersion generateNew() {
+      // Hot Rod streaming cache expects a version higher than zero.
+      long version;
+      do {
+         version = Math.abs(ThreadLocalRandom.current().nextLong());
+      } while (version <= 0);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/versioning/RandomVersionGenerator.java:20]: Looks like it can be removed.

**Review by wburns — CHANGES_REQUESTED** [review]

**Comment by wburns** [comment]
> I'm going to take a look at the code next. But I'm +1 on having the version take into account the node's address instead of being random.

Are we thinking just the node id and an incrementing number then? Also a question, but when you restart the cluster won't it have the same JGroups address during a stable shutdown?

File: core/src/main/java/org/infinispan/container/versioning/RandomVersionGenerator.java (hunk 1 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/RandomVersionGenerator.java:20]: ops... I added the `Math.abs` and forgot to remove the loop :facepalm:

**Review by jabolina — APPROVED** [review]

**Comment by jabolina** [comment]
Thanks, @pruivo!

**Pull request merged**